app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt

```diff
@@ -1,26 +1,77 @@
 package compiler.intermediate_form
 
+import compiler.ast.Function
+import compiler.ast.Type
 import compiler.ast.Variable
 
 data class FunctionDetailsGenerator(
     val depth: Int,
     val vars: Map<Variable, Boolean>,
-    val parameters: List<Variable>
+    val parameters: List<Variable>,
+    val functionCFG: ControlFlowGraph,
+    val function: Function
 )
+
 data class FunctionCallIntermediateForm(
     val callGraph: ControlFlowGraph,
     val result: IntermediateFormTreeNode?
 )
 
-fun generateCall(args: List<IntermediateFormTreeNode>): FunctionCallIntermediateForm {
-    return TODO()
+val FUNCTION_RESULT_REGISTER = Register()
+
+fun generateCall(fdg: FunctionDetailsGenerator, args: List<IntermediateFormTreeNode>): FunctionCallIntermediateForm {
+    val cfgBuilder = ControlFlowGraphBuilder()
+
+    var last: Pair<IFTNode, CFGLinkType>? = null
+
+    // write arg values to params
+    for ((arg, param) in args zip fdg.parameters) {
+        val node = genWrite(param, arg, true)
+        cfgBuilder.addLink(last, node)
+        last = Pair(node, CFGLinkType.UNCONDITIONAL)
+    }
+
+    // add function graph
+    cfgBuilder.addAllFrom(fdg.functionCFG)
+    cfgBuilder.addLink(last, fdg.functionCFG.entryTreeRoot!!)
+
+    fun modifyReturnNodesToStoreResultInAppropriateRegister(functionCfg: ControlFlowGraph): FunctionCallIntermediateForm {
+        val finalNodes = functionCfg.finalTreeRoots
+
+        val modifiedCfgBuilder = ControlFlowGraphBuilder()
+
+        val linksToIterateOver = hashMapOf(
+            CFGLinkType.UNCONDITIONAL to functionCfg.unconditionalLinks,
+            CFGLinkType.CONDITIONAL_TRUE to functionCfg.conditionalTrueLinks,
+            CFGLinkType.CONDITIONAL_FALSE to functionCfg.conditionalFalseLinks
+        )
+
+        for ((linkType, links) in linksToIterateOver) {
+            for ((from, to) in links) {
+                if (to !in finalNodes)
+                    modifiedCfgBuilder.addLink(Pair(from, linkType), to)
+                else {
+                    val newReturnNode = IntermediateFormTreeNode.RegisterWrite(FUNCTION_RESULT_REGISTER, to)
+                    modifiedCfgBuilder.addLink(Pair(from, linkType), newReturnNode)
+                }
+            }
+        }
```

Comment:
This should be a function of the `ControlFlowGraph`. Something like `updateFinalLinks`.

Reply:
Agree, but I'd rather keep `ControlFlowGraph` as simple `data class` with no methods, so I moved this to `ControlFlowGraphBuilder`